singularity-eos/eos/modifiers/ramps_eos.hpp

```diff
@@ -0,0 +1,213 @@
+    return t_.TemperatureFromDensityInternalEnergy(rho, sie, lambda);
+  }
+  PORTABLE_FUNCTION
+  Real InternalEnergyFromDensityTemperature(const Real rho, const Real temperature,
+                                            Real *lambda = nullptr) const {
+    return t_.InternalEnergyFromDensityTemperature(rho, temperature, lambda);
+  }
+  PORTABLE_FUNCTION
+  Real PressureFromDensityInternalEnergy(const Real rho, const Real sie,
+                                         Real *lambda = nullptr) const {
+    // ramp pressure
+    const Real p_ramp{get_ramp_pressure(rho)};
+    // pressure from eos
+    const Real p_eos{t_.PressureFromDensityInternalEnergy(rho, sie, lambda)};
+    // return max(p_ramp, p_eos)
+    return p_eos < p_ramp ? p_ramp : p_eos;
+  }
+  PORTABLE_FUNCTION
+  Real SpecificHeatFromDensityInternalEnergy(const Real rho, const Real sie,
+                                             Real *lambda = nullptr) const {
+    return t_.SpecificHeatFromDensityInternalEnergy(rho, sie, lambda);
+  }
+  PORTABLE_FUNCTION
+  Real BulkModulusFromDensityInternalEnergy(const Real rho, const Real sie,
+                                            Real *lambda = nullptr) const {
+    return t_.BulkModulusFromDensityInternalEnergy(rho, sie, lambda);
+  }
+  PORTABLE_FUNCTION
+  Real GruneisenParamFromDensityInternalEnergy(const Real rho, const Real sie,
+                                               Real *lambda = nullptr) const {
+    return t_.GruneisenParamFromDensityInternalEnergy(rho, sie, lambda);
+  }
+  PORTABLE_FUNCTION
+  Real PressureFromDensityTemperature(const Real rho, const Real temperature,
+                                      Real *lambda = nullptr) const {
+    const Real p_ramp{get_ramp_pressure(rho)};
+    const Real p_eos{t_.PressureFromDensityTemperature(rho, temperature, lambda)};
+    return p_eos < p_ramp ? p_ramp : p_eos;
+  }
+  PORTABLE_FUNCTION
+  Real SpecificHeatFromDensityTemperature(const Real rho, const Real temperature,
+                                          Real *lambda = nullptr) const {
+    return t_.SpecificHeatFromDensityTemperature(rho, temperature, lambda);
+  }
+  PORTABLE_FUNCTION
+  Real BulkModulusFromDensityTemperature(const Real rho, const Real temperature,
+                                         Real *lambda = nullptr) const {
+    return t_.BulkModulusFromDensityTemperature(rho, temperature, lambda);
+  }
+  PORTABLE_FUNCTION
+  Real GruneisenParamFromDensityTemperature(const Real rho, const Real temperature,
+                                            Real *lambda = nullptr) const {
+    return t_.GruneisenParamFromDensityTemperature(rho, temperature, lambda);
+  }
+  PORTABLE_FUNCTION
+  void FillEos(Real &rho, Real &temp, Real &energy, Real &press, Real &cv, Real &bmod,
+               const unsigned long output, Real *lambda = nullptr) const {
+    // density must be an input
+    assert(!(output & thermalqs::density));
```

Comment:
We'll need P-T initialization for this to be used in EAP. Since the ramp is bilinear, I'd think it would be pretty easy to write a function to invert it so that you can get rho(P) instead of P(rho).

The tricky part will be figuring out if you're on the ramp or not without knowledge of `r1`. Maybe you can call `FillEos` assuming no ramp and then compare the output density of that to the value you get out of the ramp and see which one is greater. If the EOS density is greater than the ramp then you're on the ramp.

So maybe something like
```c++
if (output & thermalqs:density) {
  assert(!(output & thermalqs::pressure));  // Need pressure if density isn't specified
  Real temp_in = temp; 
  Real energy_in = energy;
  Real press_in = press;
  t_.Filleos(rho, temp_in, energy_in, press_in, cv, bmod, lambda);
  const Real rho_ramp = get_ramp_density(press);
  if (rho > rho_ramp) {
    rho = rho_ramp;
  }
}
```
There may be more I'm missing though like how to modify the bulk modulus...

Reply:
> We'll need P-T initialization for this to be used in EAP. Since the ramp is bilinear, I'd think it would be pretty easy to write a function to invert it so that you can get rho(P) instead of P(rho).

@jdolence and I actually got a request from Ristra recently for a general API for this for all singularity-eos EOS's. I think it's a great opportunity for the base class. We can have a default functionality that uses a numerical root find and overload as needed.

Reply:
> > We'll need P-T initialization for this to be used in EAP. Since the ramp is bilinear, I'd think it would be pretty easy to write a function to invert it so that you can get rho(P) instead of P(rho).
> 
> @jdolence and I actually got a request from Ristra recently for a general API for this for all singularity-eos EOS's. I think it's a great opportunity for the base class. We can have a default functionality that uses a numerical root find and overload as needed.

You mean putting a general version of `FillEos` in the base class? At the very least we definitely need to standardize its behavior across all of the classes, but yes I agree that a general implementation in the base class would be a good idea. I'll create an issue for this.

But regardless, I would still think we need a special `FillEos` here since the ramp complicates things.

Reply:
> You mean putting a general version of FillEos in the base class? 

I meant providing a `rho(P, T)` and/or `rho(P, e)` function in the base class. That way it's available for `FillEos` to use for individual EOS's. But yeah I agree we need to standardize behaviour

Reply:
> > You mean putting a general version of FillEos in the base class?
> 
> I meant providing a `rho(P, T)` and/or `rho(P, e)` function in the base class. That way it's available for `FillEos` to use for individual EOS's. But yeah I agree we need to standardize behaviour

Ah okay, yes those are probably good ideas

Reply:
@jhp-lanl could you take a look at the most recent changes and see if they address your concerns?

Reply:
@jhp-lanl could you take a look at the most recent changes and see if they address your concerns?